### PR TITLE
gpt.py updated

### PR DESCRIPTION
**Pull request by shreyas301197**
1. gpt.py updated with inference and accuracy calculations
2. inferencing on 4 initial folders :
 folders = [
    "QP_Odisha NTSE Stage 1 2017-18) MAT",
    "QP_Maharashtra NTSE Stage 1 2017-18 MAT",
    "QP_Gujarat NTSE Stage 1 2017-18 MAT",
    "QP_Karnataka NTSE Stage1 (2017-18) MAT"
]

### CODE REVIEW

File: datastore/QP_Maharashtra NTSE Stage 1 2017-18 MAT/annotations.csv [hunk 1]
```diff
@@ -1,101 +1,100 @@
...
+ce48ecfa-9052-5cca-accd-6cbe128e3571,q7,2,['q7_0.png'],,[],,,3,4-Jan,
+ce48ecfa-9052-5cca-accd-6cbe128e3571,q8,2,['q8_0_c3.png'],,['c3.png'],,,1,4-Jan,
+ce48ecfa-9052-5cca-accd-6cbe128e3571,q9,2,['q9_0_c3.png'],,['c3.png'],,,4,4-Jan,
+ce48ecfa-9052-5cca-accd-6cbe128e3571,q10,3,['q10_0.png'],,[],,,1,4-Jan,
+ce48ecfa-9052-5cca-accd-6cbe128e3571,q11,3,['q11_0.png'],,[],,,2,4-Jan,
+ce48ecfa-9052-5cca-accd-6cbe128e3571,q12,3,['q12_0.png'],,[],,,2,4-Jan,
+ce48ecfa-9052-5cca-accd-6cbe128e3571,q13,3,['q13_0.png'],,[],,,1,4-Jan,
+ce48ecfa-9052-5cca-accd-6cbe128e3571,q14,3,['q14_0.png'],,[],,,3,4-Jan,
+ce48ecfa-9052-5cca-accd-6cbe128e3571,q15,3,['q15_0.png'],,[],,,2,4-Jan,
+ce48ecfa-9052-5cca-accd-6cbe128e3571,q18,4,['q18_0_c4.png'],,['c4.png'],,,3,4-Jan,
```
Comment (kevinscaria) [inline datastore/QP_Maharashtra NTSE Stage 1 2017-18 MAT/annotations.csv:17]: Is the conversion from `1-4` to `4-Jan` intended? Looks like formatting by Excel.

File: scripts/model inference scripts/gpt.py [hunk 2]
```diff
@@ -5,22 +5,186 @@
...
+            mime_type = 'application/octet-stream'  # Default MIME type if none is found
+
+        # Read and encode the image file
+        with open(image_path, "rb") as image_path:
+            base64_encoded_data = base64.b64encode(image_path.read()).decode('utf-8')
+
+        # Construct the data URL
+        return f"data:{mime_type};base64,{base64_encoded_data}"
+
+    def gen_template(self,
```
Comment (kevinscaria) [inline scripts/model inference scripts/gpt.py:97]: nit: better to rename it to more expressive names.

File: scripts/model inference scripts/gpt.py [hunk 3]
```diff
@@ -5,22 +5,186 @@
...
+                     images,
+                     model = "gpt-4-vision-preview",
```
Comment (kevinscaria) [inline scripts/model inference scripts/gpt.py:99]: can't we reference this using the instance field `self.model`?

File: scripts/model inference scripts/gpt.py [hunk 4]
```diff
@@ -5,22 +5,186 @@
...
+        payload = self.gen_template(imagelist)
+        response = self.query_openai(payload)
+        ans = response['choices'][0]['message']['content'].replace("```json","").replace("\n","").replace("```","")
+        try : 
+            ans = literal_eval(ans)
+            return ans
+        except :
+            ans += '"}'
+            print(f"Error while parsing response for {q_no}.Retrying")
+            ans = literal_eval(ans)
```
Comment (kevinscaria) [inline scripts/model inference scripts/gpt.py:145]: Are yo sure this literal_eval would work with the addition of `ans += '"}'`

File: scripts/model inference scripts/gpt.py [hunk 5]
```diff
@@ -5,22 +5,186 @@
...
+
+    def get_accuracy(self) :
+        return self.exact_matches/self.total_count
+    
+    def delete_response_csv(self,del_file):
+        if os.path.exists(self.response_csv_path_folder_map[del_file]) :
+            os.remove(self.response_csv_path_folder_map[del_file]) 
+        return
+    
+    def __call__(self):
```
Comment (kevinscaria) [inline scripts/model inference scripts/gpt.py:165]: Just a doubt: Any reason why we have the dunder `__call__` method?

**Review by kevinscaria — CHANGES_REQUESTED** [review]
General comment: I will refactor it eventually to accommodate it as a general class for all API based calls. 

Other comments are nits, not super necessary. Just ensure the formatting issue is intended before pushing.

File: datastore/QP_Maharashtra NTSE Stage 1 2017-18 MAT/annotations.csv (hunk 1 above)
Comment (shreyas301197) [reply datastore/QP_Maharashtra NTSE Stage 1 2017-18 MAT/annotations.csv:17]: discussed this with Ujjwala, not super important but would eventually be fixed in the annotations code

File: scripts/model inference scripts/gpt.py (hunk 3 above)
Comment (shreyas301197) [reply scripts/model inference scripts/gpt.py:99]: fixed

File: scripts/model inference scripts/gpt.py (hunk 4 above)
Comment (shreyas301197) [reply scripts/model inference scripts/gpt.py:145]: yes this is needed in edge cases when the response exceeds token limit.

**Pull request closed without merging**